### PR TITLE
Dw/cancel pending payment

### PR DESCRIPTION
**Pull request by daliwei**
@pierre I implemented the flow to canceling pending payment if a pre-specified time period has passed since its creation. Please review.

### CODE REVIEW

File: lib/paypal_express/api.rb [hunk 1]
```diff
@@ -83,15 +85,28 @@ def refund_payment(kb_account_id, kb_payment_id, kb_payment_transaction_id, kb_p
...
+        only_pending_transaction = t_info_plugins.find { |t_info_plugin| target_transaction_types.include?(t_info_plugin.transaction_type) && t_info_plugin.status != :PENDING }.nil?
         t_info_plugins_without_pending = t_info_plugins.reject { |t_info_plugin| target_transaction_types.include?(t_info_plugin.transaction_type) && t_info_plugin.status == :PENDING }
 
-        # If only pending transactions exist, return the pending one. Otherwise, only return those non-pending transaction.
-        is_transaction_pending ? t_info_plugins: t_info_plugins_without_pending
+        # If its token has expired, cancel the payment and update the response row.
+        if only_pending_transaction
+          return t_info_plugins unless token_expired(t_info_plugins.last)
+          if cancel_pending_transaction(t_info_plugins.last).nil?
+            logger.warning("Try to cancel pending transaction #{t_info_plugins.last.kb_transaction_payment_id} but failed")
```
Comment (pierre) [inline lib/paypal_express/api.rb]: I believe it's `logger.warn` instead.

Comment (daliwei) [inline lib/paypal_express/api.rb]: Yup. Should be logger.warn

File: lib/paypal_express/api.rb [hunk 2]
```diff
@@ -83,15 +85,28 @@ def refund_payment(kb_account_id, kb_payment_id, kb_payment_transaction_id, kb_p
...
+            t_info_plugins
+          else
+            logger.info("Cancel pending transaction #{t_info_plugins.last.kb_transaction_payment_id}")
```
Comment (pierre) [inline lib/paypal_express/api.rb]: Starting with Kill Bill 0.16.4, we switched to a Splunk-friendly message format (see http://dev.splunk.com/view/logging-best-practices/SP-CAAADP6): changing the message to `"Cancel pending kb_payment_transaction_id='#{t_info_plugins.last.kb_transaction_payment_id}'"` will let us search for `kb_payment_transaction_id` more easily.

See https://github.com/killbill/killbill-cybersource-plugin/blob/master/lib/cybersource/api.rb#L336 for example.

Comment (daliwei) [inline lib/paypal_express/api.rb]: Gotcha.

File: lib/paypal_express/models/response.rb [hunk 3]
```diff
@@ -77,6 +77,22 @@ def self.initial_payment_account_processor_id(kb_account_id, kb_tenant_id, token
         response.nil? ? nil : response.payment_processor_account_id
       end
 
+      def self.update_response_for_pending_payment(transaction_plugin_info)
+        response = where(:api_call => 'build_form_descriptor',
+                         :kb_payment_id => transaction_plugin_info.kb_payment_id,
+                         :kb_payment_transaction_id => transaction_plugin_info.kb_transaction_payment_id).last
+        return nil if response.nil?
+        response.update_pending_to_cancel
```
Comment (pierre) [inline lib/paypal_express/models/response.rb]: Could you simply do `where().update_all()`? Hopefully, a single transaction would be created in that case?

Might be relevant: http://stackoverflow.com/questions/9186822/how-do-i-do-an-update-statement-where-a-condition.

**Comment by daliwei** [comment]
@pierre Updated the codes. Please take a look.

File: lib/paypal_express/models/response.rb [hunk 4]
```diff
@@ -78,19 +78,11 @@ def self.initial_payment_account_processor_id(kb_account_id, kb_tenant_id, token
       end
 
       def self.update_response_for_pending_payment(transaction_plugin_info)
```
Comment (pierre) [inline lib/paypal_express/models/response.rb]: Maybe let's rename this to something more explicit (e.g. `cancel_pending_payment`)?

**Comment by pierre** [comment]
One small remark but 👍 otherwise!

**Comment by daliwei** [comment]
@pierre Done. Please check.

**Comment by pierre** [comment]
🏆

**Pull request merged**